src/controller/python/chip/discovery/__init__.py

```diff
@@ -0,0 +1,213 @@
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import logging
+import time
+import threading
+
+from dataclasses import dataclass
+from typing import List, Dict, Set, Callable
+
+from chip.discovery.library_handle import _GetDiscoveryLibraryHandle
+from chip.discovery.types import DiscoverSuccessCallback_t, DiscoverFailureCallback_t
+
+@dataclass(unsafe_hash=True)
+class PeerId:
+  """Represents a remote peer id."""
+  fabricId: int
+  nodeId: int
+
+@dataclass(unsafe_hash=True)
+class NodeAddress:
+  """Represents a distinct address where a node can be reached."""
+  interface: int
+  ip: str
+  port: int
+
+@dataclass
+class AggregatedDiscoveryResults:
+  """Discovery results for a node."""
+  peerId: PeerId
+  addresses: Set[NodeAddress]
+
+
+@dataclass
+class PendingDiscovery:
+  """Accumulator for ongoing discovery."""
+  result: AggregatedDiscoveryResults
+  callback: Callable[[AggregatedDiscoveryResults], None]
+  expireTime: int
+  firstResultTime: int
+
+# Milliseconds to wait for additional results onece a single result has
+# been received
+_RESULT_WAIT_TIME_SEC = 0.05
+
+class _PendingDiscoveries:
+  """Manages a list of pending discoveries and associated callbacks."""
+
+  activeDiscoveries: List[PendingDiscovery] = []
+
+  def __init__(self):
+    self.operationCondition = threading.Condition()
+    self.resolution = threading.Thread(target=self.ResolutionThread, daemon=True)
+    self.resolution.start()
+
+  def Start(self, peerId: PeerId, callback: Callable[[AggregatedDiscoveryResults], None], timeoutMs: int):
+    """Add a new pending result item to the internal list."""
+    with self.operationCondition:
+       self.activeDiscoveries.append(
+         PendingDiscovery(
+           AggregatedDiscoveryResults(peerId, addresses = set()),
+           callback = callback,
+           expireTime = time.time() + timeoutMs/1000.0,
+           firstResultTime = 0,
+       ))
+       self.operationCondition.notify()
+
+  def OnSuccess(self, peerId: PeerId, address: NodeAddress):
+    """Notify of a succesful address resolution."""
+    with self.operationCondition:
+      for item in self.activeDiscoveries:
+        if item.result.peerId != peerId:
+          continue
+      
+        item.result.addresses.add(address)
+        if item.firstResultTime == 0:
+          item.firstResultTime = time.time()
+      self.operationCondition.notify()
+      
+  
+  def ResolutionThread(self):
+    while True:
+      with self.operationCondition:
+        self.operationCondition.wait(self.ComputeNextEventTimeoutSeconds())
+
+        updatedDiscoveries = []
+
+        for item in self.activeDiscoveries:
+          if self.NeedsCallback(item):
+            try:
+              item.callback(item.result)
+            except:
```

Comment:
Please don't catch every possible exception.

Reply:
This was intentional: I had typo errors inside callbacks while working with different argument types and that deadlocked the application. Callbacks are not supposed to have exceptions however if they do, I do not want to deadlock.

Do you see a specific need here or was this a general comment against pokemon-style catching?